[이진우] sprint8

ajantang opened the pull request:
## 요구사항

### 기본
- [x] Github에 스프린트 미션 PR을 만들어 주세요.
- [x] Next.js를 사용해 진행합니다.

### 자유 게시판 페이지
- [x] 게시글 목록에서 드롭다운을 사용하여 "최신 순"으로 정렬할 수 있도록 합니다.
- [x] 본인이 이전 미션에서 생성한 게시글 목록 조회 API를 활용해 GET 메서드로 데이터를 가져옵니다.
- [x] 게시글 제목에 검색어가 일부 포함되면 해당 게시글을 검색할 수 있도록 합니다.
- [x] 이미지는 디폴트 이미지로 프론트엔드에서 처리해 주세요.
- [x] 게시글 닉네임 및 좋아요 개수 역시 임의값으로 프론트엔드에서 처리해주세요.
- [x] 베스트 게시글은 최신순 3개 게시글을 요청으로 데이터를 가져와 구현해주세요.
- [x] 자유게시판 페이지에서 특정 게시글을 클릭하면 해당 게시물의 상세 페이지로 이동합니다.

### 게시글 등록 & 수정 페이지
- [x] 각 input 필드에 정확한 placeholder 값을 입력합니다.
- [x] 모든 input 필드에 값을 입력하면 '등록' 버튼이 활성화됩니다.
- [x] 본인이 이전 미션에서 생성한 게시글 생성 API를 활용해 POST 메서드로 게시글을 등록합니다.
- [x] '등록' 버튼을 누르면 해당 게시물 상세 페이지로 이동합니다.
- [x] 게시글 수정 페이지 UI는 게시글 등록 페이지와 동일합니다.
- [x] 본인이 이전 미션에서 생성한 게시글 상세 API의 PATCH 메소드를 사용하여 게시물을 수정합니다.

### 게시글 상세 페이지
- [x] 본인이 이전 미션에서 생성한 게시글 상세 API의 GET 메소드를 사용하여 데이터를 가져옵니다.
- [x] 본인이 이전 미션에서 생성한 게시글 상세 API의 DELETE 메소드를 사용하여 게시물을 삭제합니다.
- [x] 댓글 input에 값을 입력하면 '등록' 버튼이 활성화됩니다.
- [x] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 POST 메소드로 댓글을 등록합니다.
- [ ] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 PATCH 메소드로 댓글을 수정합니다.
- [ ] 본인이 이전 미션에서 생성한 댓글 생성 API를 활용해 DELETE 메소드로 댓글을 삭제합니다.

### 심화
- [x] 디자인 시안에 따라 반응형 디자인을 구현합니다.
- [ ] (생략 가능) 원한다면 지금까지 진행한 모든 React 코드를 Next.js로 마이그레이션 해주세요.
	- [ ] 마이그레이션에 상당한 시간이 소요될 수 있으므로 진행을 권장하지 않습니다.


## 주요 변경사항

- client/server component 분할 작업 진행중

## 스크린샷

![image](https://ifh.cc/g/Z3mxlf.png)

## 멘토에게

- App router을 사용해서 구현하였습니다.
- 아직 client / server component 를 세분화하여 나누는 작업을 진행중입니다(이 작업 중에 꼬이는 문제가 많이 발생해서 댓글 수정 UI를 처리 못했습니다. API/BE는 구현되어 있습니다)
- 유저/좋아요 등 기본 요구사항이 아닌 테이블도 구현되어 있고, 데이터도 연결해두었습니다.
- 아직 다른 브랜치에 만들어 둔 코드를 다 옮겨 오지 못했습니다.
- 가능하다면, 추가로 작업을 진행해서 배포해보겠습니다.

ajantang committed:
feat : upload ver 1.0

ajantang committed:
chore : gitignore 파일 추가

react
VSCode
NextJS
dotenv

ajantang committed:
chore : install next.js

ajantang committed:
feat : CSR기반용 component & css 추가

무한 스크롤 & width에 따른 로직 변경을 위한 로직 수정

참고 : BE에 이미 추가로 개발된 테이블이 있어, 추가로 수정 중

ajantang committed:
chore : gitignore 내용 수정

react-이진우-sprint8 브랜치에서 작업중 적용한 내용 추가

ajantang committed:
feat : set article, bestpost value from APO res data

API 연결 및 적용

ajantang committed:
feat : add Input component

스타일이 유사한 input tag를 관리하기 위해 Input component 추가

ajantang committed:
feat : apply api to create article

API 설정 추가

ajantang committed:
feat : develop components for bulletin-board/[articleId] page

게시글 페이지 관련 component 개발 및 수정

ajantang committed:
feat : add lucon library

Date에서 client의 시간에 맞춰 값을 변경하기 위해 설치

ajantang committed:
feat : arrange code to apply design

요구사항에 맞춰 출력하기 위해 css와 일부 로직 수정

ajantang committed:
feat : apply api in /bulletin-board/[articleId] page

article, article-comment 영역에 api 적용

ajantang committed:
feat : comment maker init state after regist comment

댓글 등록 후 init value/state of comment maker
등록 댓글 화면에 출력

wildCodingWarrior commented on tailwind.config.js:
```diff
@@ -0,0 +1,32 @@
+    "./components/**/*.{js,ts,jsx,tsx,mdx}",
+    "./app/**/*.{js,ts,jsx,tsx,mdx}",
+  ],
+  theme: {
+    extend: {
+      screens: {
+        pc: 1200,
+        ta: 744,
+        mo: 375,
+      },
+      colors: {
+        background: "var(--background)",
+        foreground: "var(--foreground)",
+      },
+      width: {
+        "bestpost-pc": "38.4rem",
+        "bestpost-ta": "34rem",
+        "bestport-mo": "34.3rem",
+      },
+      height: {
+        "bestpost-pc": "16.9rem",
+        "bestpost-ta": "19.8rem",
+        "bestport-mo": "19.8rem",
+      },
+    },
+  },
```
tailwind 사용한 것 좋네요.

ajantang replied:
일단 시도는 해봤는데, 결국은 tailwind를 많이는 못 쓴거 같습니다.

wildCodingWarrior commented on lib/axios.js:
```diff
@@ -0,0 +1,102 @@
+  } catch (err) {
+    alert(err);
+  }
+}
```
API 따로 나눠서 관리하는 방식 좋습니다. 여기서 더 세분화 한다면 domain 별로 구분할 수 있겠네요. base 따로 articles 따로 comments 따로 이런식으로요.

ajantang replied:
알겠습니다. 완전히 도메인으로는 나누지 않더라도, 게시글/상품/게시글 댓글/상품 댓글 이런 식으로 나눠보겠습니다.

wildCodingWarrior commented on lib/axios.js:
```diff
@@ -0,0 +1,102 @@
+  }
+}
+
+export async function setArticle(articleId, title, content) {
```
input으로 전달되는 parameter 값은 가급적이면 RORO 형식을 따르는게 좋습니다. 

https://joojae.com/roro-pattern/

ajantang replied:
이건 생각 못해봤네요. 
일부의 속성들만 patch 로 전달된다면 RORO로 하는 것이 좋아 보입니다.

wildCodingWarrior commented on app/layout.js:
```diff
@@ -0,0 +1,43 @@
+import localFont from "next/font/local";
+import "./globals.css";
+
+const geistSans = localFont({
+  src: "./fonts/GeistVF.woff",
+  variable: "--font-geist-sans",
+  weight: "100 900",
+});
+const geistMono = localFont({
+  src: "./fonts/GeistMonoVF.woff",
+  variable: "--font-geist-mono",
+  weight: "100 900",
+});
```
이거 기본적으로 추가되는 local font 값 맞나요? 맞다면 필요없는 부분이니만큼 지우는게 더 낫겠습니다.

ajantang replied:
폰트를 다운받아서 사용하는걸 테스트 하는 과정에서 남은 코드인지, 아니면 기본적으로 생성된 코드인지 기억이 잘 안나네요. 지우도록 하겠습니다.

ajantang replied:
완

wildCodingWarrior commented on app/globals.css:
```diff
@@ -0,0 +1,207 @@
+  line-height: 1.5em;
+}
+
+.Text-md {
+  font-family: "Pretendard";
+  font-size: 1.4rem;
+  /* line-height: 24px; */
+  line-height: 1.71428em;
+}
+
+.Text-sm {
+  font-family: "Pretendard";
+  font-size: 1.3rem;
+  /* line-height: 22px; */
+  line-height: 1.6923em;
+}
+
+.Text-xs-line-height20 {
+  font-family: "Pretendard";
+  font-size: 1.2rem;
+  /* line-height: 20px; */
+  line-height: 1.66666em;
+}
+
+.Text-xs-line-height18 {
+  font-family: "Pretendard";
+  font-size: 1.2rem;
+  /* line-height: 18px; */
+  line-height: 1.5em;
+}
+
```
특별한 이유가 있는게 아니라면 kebab-case로 진행하는게 낫지 않을까요?

ajantang replied:
기존에 미션을 진행하면서 사용했던 global.css를 가져오면서 추가된 코드인데, Text로 시작하는 css는 사용하지 않고 있습니다. 이 부분도 삭제해두겠습니다.

ajantang replied:
보류 : 추가로 기존 미션 코드를 마이그레이션 하고 삭제해야 마이그레이션 과정이 수월할 것으로 판단되어, 삭제는 보류중입니다.

wildCodingWarrior commented on app/bulletin-board/page.js:
```diff
@@ -0,0 +1,20 @@
+"use client";
+
+import { DeviceProvider } from "../components/DeviceProvider";
+import BestBoard from "./BestBoard";
+import Board from "./Board";
+
+import style from "./bulletin-board.module.css";
+
+export function BulletinBoard() {
+  return (
+    <div className={style.main}>
+      <DeviceProvider>
+        <BestBoard />
+        <Board />
+      </DeviceProvider>
+    </div>
+  );
+}
+
+export default BulletinBoard;
```
해당 페이지를 use client를 하면 하위 모든 컴포넌트가 use client를 명시하지 않아도 전부 클라이언트 컴포넌트로 변하지 않나요?

ajantang replied:
맞습니다. 저 부분은 확실히 문제가 있어서 가급적 최하위 컴포넌트 위주로 client component를 만들도록 수정 중입니다.

wildCodingWarrior commented on app/bulletin-board/Board.js:
```diff
@@ -0,0 +1,161 @@
+      });
+      setList(newList);
+    });
+  };
```
이건 쉽게 하나로 만들 수 있지 않을까요?

ajantang replied:
함수로 만들어서 위로 빼두겠습니다.

ajantang replied:
완

wildCodingWarrior commented on app/bulletin-board/[articleId]/Article.js:
```diff
@@ -0,0 +1,62 @@
+  const dividerClass = `flex flex-row items-center justify-center ${style.divider}`;
+  const contentClass = `font-normal ${style.content}`;
+
+  const Owner = () => {
+    return <p className={middleBarOwnerClass}>{ownerName}</p>;
+  };
+
+  const Divider = () => {
+    return (
+      <div className={dividerClass}>
+        <Image src="/images/divider_h40.svg" width={2} height={34} alt="분리" />
+      </div>
+    );
+  };
+
```
컴포넌트를 만드실 때 가급적이면 파일을 따로 만들지 않더라도 컴포넌트 밖에서 만드는게 좋습니다.

ajantang replied:
거의 종속된 컴포넌트라 내부에 만들었는데, 컴포넌트로 따로 만들 생각은 없었으나, 형식 자체는 컴포넌트여서 내부에 만들되 대문자로 시작하는 이름으로 정의했습니다. 이부분은 이름을 수정하던지, 밖으로 옮기도록 하겠습니다.

wildCodingWarrior replied:
앗 아니요. 같은 파일에 하셔도 되지만, 컴포넌트 내부에서 컴포넌트를 정의할 필요는 없다는 말입니다.

ajantang replied:
완

wildCodingWarrior commented on .env.development:
```diff
@@ -0,0 +1,3 @@
+NEXT_PUBLIC_BASE_URL = http://localhost
+NEXT_PUBLIC_TIMEOUT = 10000
+NEXT_PUBLIC_PORT = 3001
```
나중에는 env도 gitignore에 넣으면 되겠습니다. 지금은 그럴 필요없습니다.

ajantang replied:
테스트 과정에서 필요하시다면, 이 부분은 따로 처리해두겠습니다.

ajantang replied:
완

wildCodingWarrior reviewed (commented):
고생하셨습니다!

딱히 지금 코드만 가지고는 CSR 고려할 것이 없어보이긴 합니다!

ajantang committed:
feat : apply myFavorite info & apply design bulletin-board page

자신의 favorite 선택 정보 적용
자유게시판 요구사항에 맞춰 디자인 수정

ajantang committed:
feat : arrange components

요구사항에 맞춰 컴포넌트 디자인 수정

참고 : 수정 페이지 로직 수정 필요

ajantang committed:
feat : add article modify/delete UI

게시글 수정 삭제 UI 추가
수정 삭제 로직 추가

ajantang committed:
fix : scroll function code error

언제인지 모르겠지만, 코드 수정 중 함수 로직 누락 확인. 추가

ajantang committed:
style : css setting

베스트 게시글이 반응형으로 동작할 때 늦게 반영되는 현상 임시로 처리

wildCodingWarrior commented on app/article-edit/[articleId]/articleEdit.js:
```diff
@@ -0,0 +1,180 @@
+import Input from "@/app/components/input";
+import TextArea from "@/app/components/TextArea";
+
+import {
+  MIN_TITLE_LENGTH,
+  WARN_MIN_TITLE_LENGTH,
+  MAX_TITLE_LENGHT,
+  WARN_MAX_TITLE_LENGHT,
+  MIN_CONTENT_LENGTH,
+  WARN_MIN_CONTENT_LENGTH,
+  MAX_CONTENT_LENGHT,
+  WARN_MAX_CONTENT_LENGHT,
+  VALID_VALUE,
+} from "@/app/constants/article";
```
아주 좋은 습관이네요. 정말 귀찮지만..

wildCodingWarrior commented on app/article-edit/[articleId]/articleEdit.js:
```diff
@@ -0,0 +1,180 @@
+  const [registBtnDisable, setRegistBtnDisable] = useState(false);
+  const [newTitle, setNewTitle] = useState(data.title);
+  const [content, setContent] = useState(data.content);
+  const registBtnClass = `${style["btn-regist"]}`;
+  const topBarClass = `flex flex-row items-center justify-between ${style["top-bar"]}`;
+  const topBarTextClass = `font-bold ${style["top-bar-text"]}`;
```
이렇게 사용하는 것도 좋지만, 차라리 classnames같은 라이브러리를 사용하고 해당 부분을 jsx 에서 처리하면 좀 더 관심사 분리가 명확할지도 모르겠네요,

ajantang replied:
라이브러리는 생각을 못 해봤네요. 사용 가능한 라이브러리를 찾아서 적용해보겠습니다. 저도 저런 식으로 정의하는 것이 깔끔하지도 않아서 신경이 쓰였습니다.

wildCodingWarrior commented on app/article-edit/[articleId]/articleEdit.js:
```diff
@@ -0,0 +1,180 @@
+
+    setRegistBtnDisable(false);
+
+    setArticle(articleId, newTitle, content).then((data) => {
+      const path = `/bulletin-board/${data.id}`;
+      router.push(path);
+    });
+  };
```
보통 비동기 처리를 할 때 promise를 이용해서 .then을 사용하시는 분들이 있고, async await을 선호하시는 분들도 있는데, 둘 다 괜찮으니 둘 다 익숙해져서 추후 협업시에는 팀에서 자주 사용하는 방식에 익숙해지시면 좋겠네요.

ajantang replied:
알겠습니다. 저도 배우는 과정에서 async await과 then을 사용하는 방식 다 배웠지만, then을 더 선호해서 쓰고 있습니다. 다만 promise.all 과 같이 아직 많이 써보지 않은 코드들이 많아 다양하게 시도해보겠습니다.(미션에서는 아마 then 위주로 사용하고 미션 외에 다른 코드 작성에서 then 이외의 방식을 사용해보겠습니다.)

wildCodingWarrior commented on app/article-edit/[articleId]/page.js:
```diff
@@ -0,0 +1,21 @@
+  return await data;
+}
+
+export async function ArticleEditPage({ params }) {
+  const { articleId } = params;
+
+  const data = await getArticleData(articleId);
+
+  return (
+    <div>
+      <ArticleEdit articleId={articleId} data={data} />
+    </div>
+  );
+}
+
+export default ArticleEditPage;
```
서버 컴포넌트를 이렇게 사용하는 것도 괜찮네요.

wildCodingWarrior commented on app/article-registration/articleRegistration.js:
```diff
@@ -40,7 +39,8 @@ export function ArticleRegistration() {
     }
 
-    // 등록 api 호출 + 페이지 이동
+    setRegistBtnDisable(true);
+
```
네이밍에 생각해볼 필요가 있습니다.. 실무에서는 'disabled' 같은 상태 중심의 네이밍보다는 'isValid', 'canSubmit' 등 검증(validation)이나 기능의 가용성을 나타내는 워딩을 더 선호하는 경향이 있습니다.

ajantang replied:
disabled 값으로 들어가는 변수라 일단 registBtnDisable로 변수명을 지었는데, 이 부분은 valid가 들어가는 변수명으로 바꾸도록 하겠습니다.

wildCodingWarrior commented on app/hooks/useCurrentArticleInfo.js:
```diff
@@ -0,0 +1,12 @@
+import { create } from "zustand";
+
+export const useCurrentArticleInfo = create((set) => ({
+  id: "articleId",
+  title: "article-title",
+  content: "article-content",
+  setArticleId: (newArticleId) => set({ id: newArticleId }),
+  setTitle: (newArticleTitle) => set({ title: newArticleTitle }),
+  setContent: (newArticleContent) => set({ content: newArticleContent }),
+}));
+
+export default useCurrentArticleInfo;
```
새로고침해도 정보가 남아있길 바란다면 persist 추가해주면 좋겠네요.

wildCodingWarrior reviewed (approved):
고친 부분 적절히 잘 반영하신 것 같습니다. 

가볍게 코멘트 남겼습니다.

ajantang committed:
refactor(mentor) : 피드백 내용 반영

컴포넌트 내부에 컴포넌트 선언 지양
반복 부분 함수화

ajantang committed:
feat : install autoprefixer, classnames lib

autoprefixer > 기본 브라우저 호환성을 위해 권장됨
classnaems > tailwind + @ 로 class를 정의할 때 깔끔하게 정리
